x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.enrollment;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.ssl.SslConfigException;
+import org.elasticsearch.common.ssl.SslUtil;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.security.tool.CommandLineHttpClient;
+import org.elasticsearch.xpack.security.tool.HttpResponse;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CreateEnrollmentToken {
+    protected static final String ENROLL_API_KEY_EXPIRATION = "30m";
+
+    private static final Logger logger = LogManager.getLogger(CreateEnrollmentToken.class);
+    private final Environment environment;
+    private final CommandLineHttpClient client;
+
+    public CreateEnrollmentToken(Environment environment) {
+        this(environment, new CommandLineHttpClient(environment));
+    }
+
+    // protected for testing
+    protected CreateEnrollmentToken(Environment environment, CommandLineHttpClient client) {
+        this.environment = environment;
+        this.client = client;
+    }
+
+    public String create(String user, SecureString password, SecureString keystore_password) throws Exception {
+        if (XPackSettings.ENROLLMENT_ENABLED.get(environment.settings()) != true) {
```

Comment:
No, we haven't discussed it, just haven't seen this option

Reply:
>You might have discussed this in the previous PR while I was away, but what is our reasoning around executing the calls via the API with user credentials instead of using an origin as one of the internal users ?

The origin privilege is only available if the code runs on server, which this code doesn't.